terminal-view/src/main/java/com/termux/view/textselection/TextSelectionCursorController.java

```diff
@@ -0,0 +1,382 @@
+                        terminalView.mTermSession.clipboardText(selectedText);
+                        terminalView.stopTextSelectionMode();
+                        break;
+                    case ACTION_PASTE:
+                        terminalView.stopTextSelectionMode();
+                        ClipboardManager clipboard = (ClipboardManager) terminalView.getContext().getSystemService(Context.CLIPBOARD_SERVICE);
+                        ClipData clipData = clipboard.getPrimaryClip();
+                        if (clipData != null) {
+                            CharSequence paste = clipData.getItemAt(0).coerceToText(terminalView.getContext());
+                            if (!TextUtils.isEmpty(paste)) terminalView.mEmulator.paste(paste.toString());
+                        }
+                        break;
+                    case ACTION_MORE:
+                        terminalView.stopTextSelectionMode(); //we stop text selection first, otherwise handles will show above popup
+                        terminalView.showContextMenu();
+                        break;
+                }
+
+                return true;
+            }
+
+            @Override
+            public void onDestroyActionMode(ActionMode mode) {
+            }
+
+        };
+
+        mActionMode = terminalView.startActionMode(new ActionMode.Callback2() {
+            @Override
+            public boolean onCreateActionMode(ActionMode mode, Menu menu) {
+                return callback.onCreateActionMode(mode, menu);
+            }
+
+            @Override
+            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
+                return false;
+            }
+
+            @Override
+            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
+                return callback.onActionItemClicked(mode, item);
+            }
+
+            @Override
+            public void onDestroyActionMode(ActionMode mode) {
+                // Ignore.
+            }
+
+            @Override
+            public void onGetContentRect(ActionMode mode, View view, Rect outRect) {
+                int x1 = Math.round(mSelX1 * terminalView.mRenderer.getFontWidth());
+                int x2 = Math.round(mSelX2 * terminalView.mRenderer.getFontWidth());
+                int y1 = Math.round((mSelY1 - 1 - terminalView.getTopRow()) * terminalView.mRenderer.getFontLineSpacing());
+                int y2 = Math.round((mSelY2 + 1 - terminalView.getTopRow()) * terminalView.mRenderer.getFontLineSpacing());
+
+
+                if (x1 > x2) {
+                    int tmp = x1;
+                    x1 = x2;
+                    x2 = tmp;
+                }
+
+                outRect.set(x1, y1 + mHandleHeight, x2, y2 + mHandleHeight);
+            }
+        }, ActionMode.TYPE_FLOATING);
+    }
+
+    @Override
+    public void updatePosition(TextSelectionHandleView handle, int x, int y) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.